### ISSUE TITLE
RLS runs with the home directory as the CWD, breaking RLS wrappers

### ISSUE THREAD

**Issue by joepie91** [rust-lang/rls-vscode#448]
When RLS is started by this plugin, it's run with the home directory as the current working directory, as opposed to the crate root. This breaks wrapper scripts that rely on project-specific information to run RLS in a particular environment or with a particular configuration.

For example, this is what I use for running RLS in the context of a Nix shell (essentially a self-contained environment with all the project dependencies present), under [`ide-rust`](https://github.com/rust-lang-nursery/atom-ide-rust) in Atom:

```sh
if [ -e shell.nix ]
then
    nix-shell --run "rls $@"
else
    rls $@
fi
```

That script is then configured as the RLS path, and that will result in RLS being run within the correct environment. The same script does not work with this plugin, since the current working directory is always `/home/sven` for me no matter the project, and therefore the build gets executed in the wrong environment (and subsequently fails).

Changing this to always pick the crate root as the cwd would resolve this issue, and I don't think it would break anything else. I'd also argue that it'd make more sense semantically, since that's how one would run it from a shell.

**Comment by nrc** [comment]
How does compilation rely on the CWD? Is it build scripts relying on it? In general, this should not be relied on because cargo is not guaranteed to be run in the project root directory (e.g., by using the `--manifest-path` option).

I think you should be able to work around this by running vscode from the CWD you want to use.

**Comment by joepie91** [comment]
While this would more generally fit any kind of wrapper that depends on project-specific settings, in my specific case it uses `nix-shell` with a `shell.nix` (essentially an environment configuration file) that exists at the root of the project directory. Think something like a `Vagrantfile`, but without involving VMs.

For RLS to be run within that `nix-shell` (which is the only way it can have access to the non-Rust project dependencies, that are needed for compilation), there has to be some way to intercept the call to RLS and instantiate it within the `nix-shell` subshell. The simplest solution for that is to use a wrapper script like above that infers the relevant `shell.nix` from the working directory, but that only works if the working directory equals the crate root.

Running VS Code from a specific CWD wouldn't work, because the `shell.nix` files are project-specific; that means I would be unable to have two different Rust projects open at the same time and have RLS work in both, since I can't specify the CWD in which a new project window is opened.

**Comment by Xanewok** [comment]
Fixed by #514.

**Comment by LukaHorvat** [comment]
Is it really fixed? This is still happening for me. Is the fix released?
